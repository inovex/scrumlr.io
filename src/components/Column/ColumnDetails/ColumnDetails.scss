@import "constants/style";

$description-line-height: $line-height--medium;
$description-target-height: 2 * $description-line-height;

$icon-size: $icon--large;

.column-details {
  display: flex;
  flex-direction: column;
  gap: $spacing--sm;
}

.column-details__name-and-settings-wrapper {
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  align-items: center;
}

.column-details__name-wrapper {
  display: flex;
  flex-direction: row;
  flex-grow: 1;

  overflow: hidden;
}

.column-details__name {
  all: unset; // reset input properties

  align-self: flex-start; // only take as much space as needed

  padding: 0 $spacing--xxs $spacing--xxs 0; // extend border space to bottom, and also slightly to the right

  font-size: $text--lg;
  font-weight: 700;

  color: $navy--900;

  border-bottom: 3px solid var(--accent-color--light);

  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;

  &--editing {
    flex-grow: 1; // in edit mode, actually do take full space
    border-bottom-style: dashed;
  }
}

.column-details__notes-count {
  color: $gray--800;
  font-size: $text--base;
  font-weight: 600;
}

.column-details__settings-icon-container {
  all: unset;

  width: $icon-size;
  height: $icon-size;

  cursor: pointer;
}
.column-details__settings-icon {
  width: $icon-size;
  height: $icon-size;

  transform: rotate(90deg);
}

.column-details__description-wrapper {
  position: relative;
  display: flex;
  flex-direction: row;
  justify-content: space-between;

  min-height: $description-target-height;

  &--edit {
    flex-direction: column;
  }
}

.column-details__description {
  flex-grow: 1;
  color: $navy--900;
  font-size: $text--base;
  font-weight: 600;
  line-height: $description-line-height;
  max-height: $description-target-height;
  overflow: hidden;

  &--expanded {
    max-height: max-content;
  }

  &--placeholder {
    color: $gray--800;
  }
}

.column-details__description-expand-icon-container {
  all: unset;
  align-self: flex-end;

  width: $icon-size;
  height: $icon-size;

  cursor: pointer;
}

.column-details__description-expand-icon {
  width: $icon-size;
  height: $icon-size;

  &--expanded {
    transform: rotate(180deg);
  }
}

.column-details__description-mini-menu {
  align-self: flex-end;
}

[theme="dark"] {
  .column-details__name {
    color: $gray--000;
    border-bottom-color: var(--accent-color--dark);
  }

  .column-details__description {
    &--placeholder {
      color: $gray--700;
    }
  }

  .column-details__settings-icon-container {
    color: $gray--000;
  }

  .column-details__description-expand-icon {
    color: rgba($gray--000, 0.7);
  }
}
