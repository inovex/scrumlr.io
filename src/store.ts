import { createStore, compose } from 'redux';
import { reactReduxFirebase } from 'react-redux-firebase';

import rootReducer from './reducers';
import { firebase as fbConfig } from './config';
import * as firebase from 'firebase';

export default function configureStore(initialState: {}, history?: {}) {
  firebase.initializeApp(fbConfig);

  const composeEnhancers =
    (window as any).__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;
  const createStoreWithMiddleware = composeEnhancers(
    reactReduxFirebase(firebase, {
      userProfile: 'users',
      enableLogging: false
    })
  )(createStore);
  const store = createStoreWithMiddleware(rootReducer);

  // tslint:disable
  if ((module as any).hot) {
    // Enable Webpack hot module replacement for reducers
    (module as any).hot.accept('./reducers', () => {
      const nextRootReducer = require('./reducers');
      store.replaceReducer(nextRootReducer);
    });
  }
  // tslint:enable

  return store;
}
