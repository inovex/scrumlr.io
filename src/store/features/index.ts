export * from "./auth";
export * from "./board";
export * from "./boardReactions";
export * from "./columns";
export * from "./notes";
export * from "./participants";
export * from "./reactions";
export * from "./requests";
export * from "./skinTone";
export * from "./view";
export * from "./votes";
export * from "./votings";
export * from "./templates";
export * from "./templateColumns";
