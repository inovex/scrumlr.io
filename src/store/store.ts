import { createStore, applyMiddleware } from 'redux';
import  rootReducer from './rootReducer';
import thunk from 'redux-thunk';
import { composeWithDevTools } from 'redux-devtools-extension';

const store = createStore(rootReducer, composeWithDevTools(applyMiddleware(thunk)));

export default store;