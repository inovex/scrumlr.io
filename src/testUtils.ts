export type AssertTypeEqual<T, Expected> =
    T extends Expected
        ? (Expected extends T ? true : never)
        : never;