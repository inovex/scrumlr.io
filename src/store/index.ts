import {createStore, applyMiddleware, combineReducers, Dispatch, MiddlewareAPI, AnyAction} from "redux";
import thunk from "redux-thunk";
import {composeWithDevTools} from "redux-devtools-extension";
import {ApplicationState} from "../types/store";
import {ReduxAction} from "./action";
import {boardReducer} from "./reducer/board";
import {noteReducer} from "./reducer/note";
import {usersReducer} from "./reducer/users";
import {passNoteMiddleware} from "./middleware/note";
import {passBoardMiddleware} from "./middleware/board";
import {passBoardJoinConfirmationMiddleware} from "./middleware/boardJoinConfirmation";
import {passColumnMiddleware} from "./middleware/column";
import {passJoinRequestMiddleware} from "./middleware/joinRequest";
import {joinRequestReducer} from "./reducer/joinRequest";

const parseMiddleware = (stateAPI: MiddlewareAPI<Dispatch<AnyAction>, ApplicationState>) => (dispatch: Dispatch) => (action: ReduxAction) => {
  try {
    return dispatch(action);
  } finally {
    passBoardJoinConfirmationMiddleware(stateAPI, dispatch, action);
    passBoardMiddleware(stateAPI, dispatch, action);
    passColumnMiddleware(stateAPI, dispatch, action);
    passNoteMiddleware(stateAPI, dispatch, action);
    passJoinRequestMiddleware(stateAPI, dispatch, action);
  }
};

const rootReducer = combineReducers<ApplicationState>({
  board: boardReducer,
  notes: noteReducer,
  users: usersReducer,
  joinRequests: joinRequestReducer,
});

const store = createStore(rootReducer, composeWithDevTools(applyMiddleware(thunk), applyMiddleware(parseMiddleware)));

export default store;
