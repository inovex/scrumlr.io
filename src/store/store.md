# Redux Store Guide

## Table of Contents

1. [Introduction](#introduction)
2. [Core Concepts](#core-concepts)
   1. [Store](#store)
   2. [Actions](#actions)
   3. [Reducers](#reducers)
   4. [Async Thunks](#async-thunks)
3. [Scrumlr Specifics](#scrumlr-specifics)

## Introduction

Since updating redux to 2.x, the whole store had to be refactored.
So, I'd like to use the opportunity to add this little writeup on how exactly the parts play together, in order to make it easier for newer devs to understand.

### Definition

Redux is a predictable state container for JavaScript apps. It helps manage the state of the application in a single, centralized store.

### Structure

The structure of the store is defined in such a way that each feature gets their own subdirectory with all the different files,
representing one required part of a redux feature.

```
store.ts
features/
├── featureA/
│   ├── actions.ts
│   ├── reducer.ts
│   ├── thunks.ts
│   └── types.ts
└── featureB/
    ├── actions.ts
    ├── reducer.ts
    ├── thunks.ts
    └── types.ts
```

## Core Concepts

### Store

The store holds the entire state tree of the application.
It allows access to the current state and dispatches actions to modify the state.

#### Retrieving the current state

In order to access the current state, a hook is used.
We can get that value using the following code throughout the project:

```tsx
import {useAppSelector} from "store";

export const CustomComponent = () => {
  const valueA = useAppSelector((state) => state.featureA.valueA);

  return <div>Value A: {valueA}</div>;
};
```

The value is automatically updated whenever it changes in the store, causing a rerender.
It is immutable, so to change the value, an _Action_ has to be _dispatched_ first.

### Actions

Actions are plain objects describing what happened.
They consist of a `type` and an optional `payload`.
Generally, they are defined like this:

```ts
import {createAction} from "@reduxjs/toolkit";

export const exampleAction = createAction<T>("featureA/actionA");
```

where `featureA/actionA` is the name (type) and `T` is the payload type.

Let's define a simple action:

```ts
export const setValueA = createAction<number>("featureA/setValueA");
```

Actions can be dispatched.

```ts
import {useAppDispatch} from "store";

const dispatch = useAppDispatch();

dispatch(setValueA(42));
```

### Reducers

Reducers specify how the state changes in response to actions.
They are pure functions that return the next state.
There are different patterns to define reducers, we're using `builder`.

Assuming the state for `reducerA` is defined like this in `types.ts`:

```ts
export type ReducerAState = {
  valueA: number;
  arrayA: string[];
};
```

First, we define the initial state:

```ts
const initialState: ReducerAState = {valueA: 0, arrayA: []};
```

the reducer for feature A would look like this:

```ts
export const featureAReducer = createReducer(initialState, (builder) =>
  builder
    .addCase(setValueA, (state, action) => {
      state.valueA = action.payload;
    })
    .addCase(addItemWay1, (state, action) => {
      state.arrayA.push(action.payload);
    })
    .addCase(addItemWay2, (state, action) => {
      return {
        ...state,
        arrayA: [...state.arrayA, action.payload],
      };
    })
);
```

`action.payload` will automatically infer the type which was defined in the `createAction` function.
As you can see, there are multiple ways to change the state.

1. Since Redux uses [Immer](https://immerjs.github.io/immer/) internally, the state inside reducers are a writable draft.
   Thus, it is actually possible to mutate state values.
2. Alternatively, a new state may be returned directly, which is sometimes more convenient. Don't forget the other values though!

**Important**: Do not rewrite the state directly!

```ts
builder.addCase(illegalRewrite, (state, action) => {
  state = action.payload; // don't do this
});
```

This changes the reference to `state`, which doesn't lead to a correct state change.

### Async Thunks

Thunks allow for asynchronous operations by wrapping expressions into a function. See https://daveceddia.com/what-is-a-thunk/.
Because Actions are just plain objects, and reducers are pure functions, thunks allow for asynchronous code and side effects.
They are defined like this:

```ts
export const myAsyncThunk = createAsyncThunk<R, T, {state: ApplicationState}>("featureA/asyncTask", async (payload, {dispatch, getState}) => {
  const result = await API.doStuff(payload);
  dispatch(someAction(result));
});
```

There's some things to unpack here:

- `R`: return type of the thunk, if not needed just put `void`
- `T`: payload type, analog to normal actions (`void` if none)
- `{state: ApplicationState}`: tells the TS what type to infer for the store

Inside the function body, any code may be called, including API calls and other dispatch calls.
For this, `dispatch` and `getState` are can optionally be passed as parameters to the function.

It is also possible to use the return value of the thunk inside a reducer:

```ts
builder
  // pending or fulfilled
  .addCase(myAsyncThunk.fulfilled, (state, action) => {
    state.someProperty = action.payload; // make sure the thunk returns the right value
  });
```

A thunk is called like any other action:

```ts
dispatch(myAsyncThunk(someParameter));
```

## Scrumlr Specifics

For Scrumlr, you should know about some extra things:

### The `retryable` function

This function abstracts error handling and retry logic for thunks.
So for example, the above custom thunk can be wrapped:

```ts
import {retryable} from "store";

export const myAsyncThunk = createAsyncThunk<
  R,
  T,
  {
    state: ApplicationState;
  }
>("featureA/asyncTask", async (payload, {dispatch, getState}) => {
  await retryable(
    () => API.doStuff(payload),
    dispatch,
    () => myAsyncThunk(payload),
    "asyncError"
  );
});
```

Since the promise is returned, it can also be further chained using `.then()` and `.catch()`.

Maybe in the future, this could also be done using a custom middleware; feel free to do so.

### Sockets and Backend

The Scrumlr Backend uses sockets for primary communication.
For that reason, there are two socket connections:

1. **requests/thunks**: Used for initial login; closed after.
2. **board/thunks**: Used for communication with the backend concerning all changes to the board, like participants, notes, etc.

For most cases, like notes, there are two separate actions for doing a task that requires the backend to change something that affects all participants.
Adding a note flows like this:

1. user clicks a button to add a note
2. dispatch thunk `addNote`
3. thunk calls backend
4. backend adds the note to the database
5. backend emits `"NOTE_ADDED"` event to the socket
6. frontend receives `"NOTE_ADDED"` event
7. it then dispatches action `addedNote`
8. reducer catches action and adds note to `NoteState`
9. state is updated
10. components reading notes using `useAppSelector` are re-rendered
