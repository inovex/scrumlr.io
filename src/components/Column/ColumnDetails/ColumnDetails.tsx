import {Column, createColumn, deleteColumnOptimistically, editColumn} from "store/features";
import {useTranslation} from "react-i18next";
import {useEffect, useRef, useState} from "react";
import classNames from "classnames";
import {ReactComponent as ArrowIcon} from "assets/icons/arrow-down.svg";
import {ReactComponent as SettingsIcon} from "assets/icons/three-dots.svg";
import {ReactComponent as CheckDoneIcon} from "assets/icons/check-done.svg";
import {ReactComponent as CloseIcon} from "assets/icons/close.svg";
import {useTextOverflow} from "utils/hooks/useTextOverflow";
import {ColumnSettings} from "components/Column/ColumnSettings";
import {TextArea} from "components/TextArea/TextArea";
import {MiniMenu, MiniMenuItem} from "components/MiniMenu/MiniMenu";
import {useAppDispatch, useAppSelector} from "store";
import {useOnBlur} from "utils/hooks/useOnBlur";
import {Tooltip} from "components/Tooltip";
import "components/Column/ColumnDetails/ColumnDetails.scss";
import {MAX_BOARD_NAME_LENGTH, MAX_COLUMN_DESCRIPTION_LENGTH} from "constants/misc";

export type ColumnDetailsMode = "view" | "edit";

export type ColumnDetailsProps = {
  column: Column;
  notesCount: number;
  mode: ColumnDetailsMode;
  changeMode: (mode: ColumnDetailsMode) => void;
  isTemporary: boolean;
};

export const ColumnDetails = (props: ColumnDetailsProps) => {
  const {t} = useTranslation();
  const dispatch = useAppDispatch();

  const isModerator = useAppSelector((state) => state.participants?.self?.role === "OWNER" || state.participants?.self?.role === "MODERATOR");

  const inputRef = useRef<HTMLInputElement>(null);

  const {isTextTruncated: isDescriptionTextTruncated, textRef: viewDescriptionRef} = useTextOverflow<HTMLTextAreaElement>(props.column.description);
  const {isTextTruncated: isNameTextTruncated, textRef: viewNameRef} = useTextOverflow<HTMLDivElement>(props.column.name);

  const [openSettings, setOpenSettings] = useState(false);
  const [isDescriptionExpanded, setIsDescriptionExpanded] = useState(false);

  const [localName, setLocalName] = useState(props.column.name);
  const [localDescription, setLocalDescription] = useState(props.column.description);

  const isValidName = localName.trim().length > 0 && localName.length <= MAX_BOARD_NAME_LENGTH;
  const isValidDescription = localDescription.length <= MAX_COLUMN_DESCRIPTION_LENGTH;
  const isValidDetails = isValidName && isValidDescription;

  // focus input upon entering edit mode
  useEffect(() => {
    if (inputRef && inputRef.current) {
      inputRef.current.focus();
    }
  }, [inputRef, props.mode]);

  const changeMode = (mode: ColumnDetailsMode) => {
    if (!isModerator) return;

    props.changeMode(mode);
  };

  const cancelUpdate = () => {
    if (props.isTemporary) {
      dispatch(deleteColumnOptimistically(props.column.id));
    } else {
      setLocalName(props.column.name);
      setLocalDescription(props.column.description);
    }
  };

  // if column exists, update details
  // else create new column
  const updateColumnDetails = (newName: string, newDescription: string) => {
    if (!isValidDetails) {
      cancelUpdate();
      return;
    }

    const updateColumnPayload: Column = {...props.column, name: newName, description: newDescription};
    if (props.isTemporary) {
      dispatch(createColumn(updateColumnPayload));
    } else {
      dispatch(editColumn({id: props.column.id, column: updateColumnPayload}));
    }

    changeMode("view");
  };

  const handleBlur = () => {
    if (props.mode === "view") return;

    // behaviour: always save as long as name is not empty.
    // could also change it to only save if persisted is empty.
    updateColumnDetails(localName, localDescription);
  };

  const ref = useOnBlur<HTMLDivElement>(handleBlur);

  const saveColumnDetailsMiniMenu: MiniMenuItem[] = [
    {
      className: "mini-menu-item--cancel",
      element: <CloseIcon />,
      label: "Cancel",
      onClick(): void {
        // reset
        cancelUpdate();
        changeMode("view");
      },
    },
    {
      className: "mini-menu-item--save",
      element: <CheckDoneIcon />,
      label: "Save",
      disabled: !isValidDetails,
      onClick(): void {
        updateColumnDetails(localName, localDescription);
        changeMode("view");
      },
    },
  ];

  const renderName = () =>
    props.mode === "view" ? (
      <>
        <div ref={viewNameRef} id={`col-${props.column.id}-name`} className="column-details__name" onDoubleClick={() => changeMode("edit")}>
          {props.column.name}
        </div>
        <div className="column-details__notes-count">{props.notesCount}</div>
        {isNameTextTruncated.horizontal && <Tooltip anchorSelect={`#col-${props.column.id}-name`} content={props.column.name} color={props.column.color} />}
      </>
    ) : (
      <input
        ref={inputRef}
        className={classNames("column-details__name", "column-details__name--editing")}
        value={localName}
        maxLength={MAX_BOARD_NAME_LENGTH}
        onInput={(e) => setLocalName(e.currentTarget.value)}
      />
    );

  const viewableDescription = () => {
    // expandable content
    if (props.column.description)
      return (
        <>
          <TextArea ref={viewDescriptionRef} input={props.column.description} embedded extendable={isDescriptionExpanded} disabled border="none" rows={2} />
          {(isDescriptionTextTruncated.vertical || isDescriptionExpanded) && (
            <button className={classNames("column-details__description-expand-icon-container")} onClick={() => setIsDescriptionExpanded((expanded) => !expanded)}>
              <ArrowIcon className={classNames("column-details__description-expand-icon", {"column-details__description-expand-icon--expanded": isDescriptionExpanded})} />
            </button>
          )}
        </>
      );
    if (isModerator) {
      // placeholder
      return <div className="column-details__description--placeholder">{t("Column.Header.descriptionPlaceholder")}</div>;
    }
    // empty placeholder space
    return <div className="column-details__description--placeholder" />;
  };

  const editableDescription = () => (
    <>
      <TextArea
        className="column-details__description-text-area"
        input={localDescription}
        setInput={setLocalDescription}
        placeholder={t("Column.Header.descriptionPlaceholder")}
        embedded
        extendable
        border="thick"
        rows={3}
        maxLength={MAX_COLUMN_DESCRIPTION_LENGTH}
      />
      <MiniMenu className="column-details__description-mini-menu" items={saveColumnDetailsMiniMenu} small transparent />
    </>
  );

  const renderDescription = () => (props.mode === "view" ? viewableDescription() : editableDescription());

  const renderSettings = () => {
    if (!isModerator) return null;

    if (openSettings) {
      return (
        <ColumnSettings
          className={classNames("column-details__settings", "column-details__settings--open")}
          column={props.column}
          onClose={() => setOpenSettings(false)}
          onNameEdit={() => changeMode("edit")}
        />
      );
    }
    return (
      <button
        className={classNames("column-details__settings", "column-details__settings--closed", "column-details__settings-icon-container")}
        onClick={() => setOpenSettings(true)}
      >
        <SettingsIcon className="column-details__settings-icon" />
      </button>
    );
  };

  return (
    <div className="column-details" ref={ref}>
      <div className="column-details__name-and-settings-wrapper">
        <div className="column-details__name-wrapper">{renderName()}</div>
        {renderSettings()}
      </div>
      <div className={classNames("column-details__description-wrapper", `column-details__description-wrapper--${props.mode}`)}>{renderDescription()}</div>
    </div>
  );
};
