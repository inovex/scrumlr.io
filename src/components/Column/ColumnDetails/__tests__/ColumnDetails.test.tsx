import {render} from "testUtils";
import {ColumnDetails, ColumnDetailsProps} from "components/Column/ColumnDetails/ColumnDetails";
import {Column, Participant, ParticipantRole} from "store/features";
import getTestApplicationState from "utils/test/getTestApplicationState";
import getTestStore from "utils/test/getTestStore";
import {Provider} from "react-redux";
import {fireEvent} from "@testing-library/react";

describe("ColumnDetails", () => {
  const renderColumnDetails = (overrideProps?: Partial<ColumnDetailsProps>, userRole: ParticipantRole = "OWNER") => {
    const defaultProps: ColumnDetailsProps = {
      column: getTestApplicationState().columns[0],
      notesCount: 1,
      mode: "view",
      isTemporary: false,
      changeMode: jest.fn(),
    };

    const props = {...defaultProps, ...overrideProps};
    const self: Participant = {...getTestApplicationState().participants.self!, role: userRole};

    return render(
      <Provider store={getTestStore({participants: {self}})}>
        <ColumnDetails {...props} />
      </Provider>
    );
  };

  it("should render correctly (view)", () => {
    const {container} = renderColumnDetails();
    expect(container).toMatchSnapshot();
  });

  it("should render correctly (edit)", () => {
    const {container} = renderColumnDetails({mode: "edit"});
    expect(container).toMatchSnapshot();
  });

  it("should switch to edit mode (self is moderator)", () => {
    const changeModeSpy = jest.fn();
    const {container} = renderColumnDetails({changeMode: changeModeSpy});
    const columnDetailNameNode = container.querySelector<HTMLDivElement>(".column-details__name")!;

    fireEvent.click(columnDetailNameNode);
    expect(changeModeSpy).toHaveBeenCalledWith("edit");
  });

  it("should not switch to edit mode (self is participant)", () => {
    const changeModeSpy = jest.fn();
    const {container} = renderColumnDetails({changeMode: changeModeSpy}, "PARTICIPANT");
    const columnDetailNameNode = container.querySelector<HTMLDivElement>(".column-details__name")!;

    fireEvent.click(columnDetailNameNode);
    expect(changeModeSpy).not.toHaveBeenCalledWith("edit");
  });

  describe("Description click editing with readOnly TextArea", () => {
    it("should switch to edit mode when clicking description placeholder (moderator)", () => {
      const changeModeSpy = jest.fn();
      const columnWithoutDescription = {...getTestApplicationState().columns[0], description: ""};
      const {container} = renderColumnDetails({column: columnWithoutDescription, changeMode: changeModeSpy});

      const placeholderNode = container.querySelector<HTMLDivElement>(".column-details__description--placeholder")!;
      expect(placeholderNode).toBeTruthy();

      fireEvent.click(placeholderNode);
      expect(changeModeSpy).toHaveBeenCalledWith("edit");
    });

    it("should not switch to edit mode when clicking description placeholder (participant)", () => {
      const changeModeSpy = jest.fn();
      const columnWithoutDescription = {...getTestApplicationState().columns[0], description: ""};
      const {container} = renderColumnDetails({column: columnWithoutDescription, changeMode: changeModeSpy}, "PARTICIPANT");

      const placeholderNode = container.querySelector<HTMLDivElement>(".column-details__description--placeholder")!;
      expect(placeholderNode).toBeTruthy();

      fireEvent.click(placeholderNode);
      expect(changeModeSpy).not.toHaveBeenCalledWith("edit");
    });

    it("should switch to edit mode when clicking filled description with readOnly TextArea (moderator)", () => {
      const changeModeSpy = jest.fn();
      const columnWithDescription = {...getTestApplicationState().columns[0], description: "Test description content"};
      const {container} = renderColumnDetails({column: columnWithDescription, changeMode: changeModeSpy});

      const textAreaWrapper = container.querySelector<HTMLDivElement>(".column-details__description-wrapper--view")!;
      expect(textAreaWrapper).toBeTruthy();

      const textArea = textAreaWrapper.querySelector<HTMLTextAreaElement>("textarea")!;
      expect(textArea).toBeTruthy();
      expect(textArea.value).toBe("Test description content");
      expect(textArea.readOnly).toBe(true);
      expect(textArea.disabled).toBe(false);

      // Click on the readOnly textarea should trigger edit mode
      fireEvent.click(textArea);
      expect(changeModeSpy).toHaveBeenCalledWith("edit");
    });

    it("should not switch to edit mode when clicking filled description (participant)", () => {
      const changeModeSpy = jest.fn();
      const columnWithDescription = {...getTestApplicationState().columns[0], description: "Test description content"};
      const {container} = renderColumnDetails({column: columnWithDescription, changeMode: changeModeSpy}, "PARTICIPANT");

      const textAreaWrapper = container.querySelector<HTMLDivElement>(".column-details__description-wrapper--view")!;
      expect(textAreaWrapper).toBeTruthy();

      const textArea = textAreaWrapper.querySelector<HTMLTextAreaElement>("textarea")!;
      expect(textArea).toBeTruthy();
      expect(textArea.readOnly).toBe(true);

      fireEvent.click(textArea);
      expect(changeModeSpy).not.toHaveBeenCalledWith("edit");
    });

    it("should focus description field when switching to edit mode via placeholder click", () => {
      const changeModeSpy = jest.fn();
      const columnWithoutDescription = {...getTestApplicationState().columns[0], description: ""};
      const {container, rerender} = renderColumnDetails({column: columnWithoutDescription, changeMode: changeModeSpy});

      const placeholderNode = container.querySelector<HTMLDivElement>(".column-details__description--placeholder")!;
      fireEvent.click(placeholderNode);
      expect(changeModeSpy).toHaveBeenCalledWith("edit");

      // Simulate mode change by re-rendering with edit mode
      const self = {...getTestApplicationState().participants.self!, role: "OWNER" as const};
      rerender(
        <Provider store={getTestStore({participants: {self}})}>
          <ColumnDetails column={columnWithoutDescription} notesCount={1} mode="edit" isTemporary={false} changeMode={changeModeSpy} />
        </Provider>
      );

      // In edit mode, the description textarea should be focusable and not readOnly
      const editTextArea = container.querySelector<HTMLTextAreaElement>(".column-details__description-text-area")!;
      expect(editTextArea).toBeTruthy();
      expect(editTextArea.readOnly).toBe(false);
      expect(editTextArea.disabled).toBe(false);
    });

    it("should focus description field when switching to edit mode via filled description click", () => {
      const changeModeSpy = jest.fn();
      const columnWithDescription = {...getTestApplicationState().columns[0], description: "Test description content"};
      const {container, rerender} = renderColumnDetails({column: columnWithDescription, changeMode: changeModeSpy});

      const textArea = container.querySelector<HTMLTextAreaElement>("textarea")!;
      fireEvent.click(textArea);
      expect(changeModeSpy).toHaveBeenCalledWith("edit");

      // Simulate mode change by re-rendering with edit mode
      const self = {...getTestApplicationState().participants.self!, role: "OWNER" as const};
      rerender(
        <Provider store={getTestStore({participants: {self}})}>
          <ColumnDetails column={columnWithDescription} notesCount={1} mode="edit" isTemporary={false} changeMode={changeModeSpy} />
        </Provider>
      );

      // In edit mode, the description textarea should be focusable and not readOnly
      const editTextArea = container.querySelector<HTMLTextAreaElement>(".column-details__description-text-area")!;
      expect(editTextArea).toBeTruthy();
      expect(editTextArea.readOnly).toBe(false);
      expect(editTextArea.disabled).toBe(false);
      expect(editTextArea.value).toBe("Test description content");
    });

    it("should show readOnly textarea is not editable but interactive", () => {
      const columnWithDescription = {...getTestApplicationState().columns[0], description: "Test description content"};
      const {container} = renderColumnDetails({column: columnWithDescription});

      const textArea = container.querySelector<HTMLTextAreaElement>("textarea")!;
      expect(textArea).toBeTruthy();
      expect(textArea.value).toBe("Test description content");
      expect(textArea.readOnly).toBe(true);
      expect(textArea.disabled).toBe(false);

      // Verify readOnly prevents typing but allows events
      expect(textArea.getAttribute("readonly")).not.toBeNull();
      expect(textArea.getAttribute("disabled")).toBeNull();
    });

    it("should show expand button for long descriptions and handle expansion", () => {
      const longDescription = "This is a very long description that should trigger the expand functionality ".repeat(10);
      const columnWithLongDescription = {...getTestApplicationState().columns[0], description: longDescription};
      const {container} = renderColumnDetails({column: columnWithLongDescription});

      const textArea = container.querySelector<HTMLTextAreaElement>("textarea")!;
      expect(textArea.value).toBe(longDescription);
      expect(textArea.readOnly).toBe(true);

      // The expand button should be present for long content
      const expandButton = container.querySelector<HTMLButtonElement>(".column-details__description-expand-icon-container");
      if (expandButton) {
        fireEvent.click(expandButton);
        // After clicking expand, the content should still be there and readOnly
        expect(textArea.value).toBe(longDescription);
        expect(textArea.readOnly).toBe(true);
      }
    });

    it("should preserve click functionality after expand/collapse", () => {
      const changeModeSpy = jest.fn();
      const longDescription = "This is a long description ".repeat(10);
      const columnWithLongDescription = {...getTestApplicationState().columns[0], description: longDescription};
      const {container} = renderColumnDetails({column: columnWithLongDescription, changeMode: changeModeSpy});

      const textArea = container.querySelector<HTMLTextAreaElement>("textarea")!;
      const expandButton = container.querySelector<HTMLButtonElement>(".column-details__description-expand-icon-container");

      // Expand the description
      if (expandButton) {
        fireEvent.click(expandButton);
      }

      // Click should still work after expansion
      fireEvent.click(textArea);
      expect(changeModeSpy).toHaveBeenCalledWith("edit");
    });
  });
});
