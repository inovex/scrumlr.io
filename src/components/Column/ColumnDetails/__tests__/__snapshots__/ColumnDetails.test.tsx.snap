// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ColumnDetails should render correctly (edit) 1`] = `
<div>
  <div
    class="column-details"
  >
    <div
      class="column-details__name-and-settings-wrapper"
    >
      <div
        class="column-details__name-wrapper"
      >
        <input
          class="column-details__name column-details__name--editing"
          maxlength="128"
          value="test-columns-name-1"
        />
      </div>
      <button
        class="column-details__settings column-details__settings--closed column-details__settings-icon-container"
      >
        <svg
          class="column-details__settings-icon"
        >
          three-dots.svg
        </svg>
      </button>
    </div>
    <div
      class="column-details__description-wrapper column-details__description-wrapper--edit"
    >
      <style>
        .text-area { --text-area-rows: 3 }
      </style>
      <textarea
        class="column-details__description-text-area text-area text-area--extendable text-area--embedded text-area--border-thick text-area--text-align-left"
        maxlength="1024"
        placeholder="Add description (optional)"
      />
      <div
        class="column-details__description-mini-menu mini-menu mini-menu--small mini-menu--transparent"
      >
        <button
          aria-label="Cancel"
          class="mini-menu-item--cancel mini-menu__item mini-menu__item--small"
          data-cy="undefined-item-Cancel"
          data-tooltip-content="Cancel"
          data-tooltip-id="scrumlr-tooltip"
          id="mini-menu-Cancel3"
        >
          <svg>
            close.svg
          </svg>
        </button>
        <button
          aria-label="Save"
          class="mini-menu-item--save mini-menu__item mini-menu__item--small"
          data-cy="undefined-item-Save"
          data-tooltip-content="Save"
          data-tooltip-id="scrumlr-tooltip"
          id="mini-menu-Save4"
        >
          <svg>
            check-done.svg
          </svg>
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`ColumnDetails should render correctly (view) 1`] = `
<div>
  <div
    class="column-details"
  >
    <div
      class="column-details__name-and-settings-wrapper"
    >
      <div
        class="column-details__name-wrapper"
      >
        <div
          class="column-details__name"
          id="col-test-columns-id-1-name"
        >
          test-columns-name-1
        </div>
        <div
          class="column-details__notes-count"
        >
          1
        </div>
      </div>
      <button
        class="column-details__settings column-details__settings--closed column-details__settings-icon-container"
      >
        <svg
          class="column-details__settings-icon"
        >
          three-dots.svg
        </svg>
      </button>
    </div>
    <div
      class="column-details__description-wrapper column-details__description-wrapper--view"
    >
      <div
        class="column-details__description--placeholder"
      >
        Add description (optional)
      </div>
    </div>
  </div>
</div>
`;
